cmd/createorupdate/update.go

```diff
@@ -0,0 +1,231 @@
+			return app.Count
+		}
+	}
+
+	panic("invalid role")
+}
+
+func listVMs(ctx context.Context, cs *api.OpenShiftManagedCluster, vmc compute.VirtualMachineScaleSetVMsClient, role api.AgentPoolProfileRole) ([]compute.VirtualMachineScaleSetVM, error) {
+	vmPages, err := vmc.List(ctx, cs.Properties.AzProfile.ResourceGroup, "ss-"+string(role), "", "", "")
+	if err != nil {
+		return nil, err
+	}
+
+	var vms []compute.VirtualMachineScaleSetVM
+	for vmPages.NotDone() {
+		vms = append(vms, vmPages.Values()...)
+
+		err = vmPages.Next()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return vms, nil
+}
+
+// updatePlusOne creates an extra VM, then runs updateInPlace, then removes the
+// extra VM.
+func updatePlusOne(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin, ssc compute.VirtualMachineScaleSetsClient, vmc compute.VirtualMachineScaleSetVMsClient, role api.AgentPoolProfileRole) error {
+	count := getCount(cs, role)
+
+	// store a list of all the VM instances now, so that if we end up creating
+	// new ones (in the crash recovery case, we might not), we can detect which
+	// they are
+	vms, err := listVMs(ctx, cs, vmc, role)
+	if err != nil {
+		return err
+	}
+
+	vmsBefore := map[string]struct{}{}
+	for _, vm := range vms {
+		vmsBefore[*vm.InstanceID] = struct{}{}
+	}
+
+	log.Infof("setting ss-%s capacity to %d", role, count+1)
+	future, err := ssc.Update(ctx, cs.Properties.AzProfile.ResourceGroup, "ss-"+string(role), compute.VirtualMachineScaleSetUpdate{
+		Sku: &compute.Sku{
+			Capacity: to.Int64Ptr(int64(count) + 1),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletion(ctx, ssc.Client)
+	if err != nil {
+		return err
+	}
+
+	vms, err = listVMs(ctx, cs, vmc, role)
+	if err != nil {
+		return err
+	}
+
+	// wait for newly created VMs to reach readiness (n.b. one alternative to
+	// this approach would be for the CSE to not return until the node is
+	// ready, but that is also problematic)
+	for _, vm := range vms {
+		if _, found := vmsBefore[*vm.InstanceID]; !found {
+			log.Infof("waiting for %s to be ready", *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+			err = p.WaitForReady(ctx, cs, role, *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	err = updateInPlace(ctx, cs, p, ssc, vmc, role)
+	if err != nil {
+		return err
+	}
+
+	// remove surplus VMs
+	for _, vm := range vms[count:] {
+		log.Infof("draining %s", *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+		err = p.Drain(ctx, cs, role, *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+		if err != nil {
+			return err
+		}
+
+		log.Infof("deleting %s", *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName)
+		future, err := vmc.Delete(ctx, cs.Properties.AzProfile.ResourceGroup, "ss-"+string(role), *vm.InstanceID)
+		if err != nil {
+			return err
+		}
+
+		err = future.WaitForCompletion(ctx, vmc.Client)
```

Comment:
This can also take some time to complete. When I was testing https://github.com/openshift/openshift-azure/pull/138, I figured out we could spend ~20m for a VM deletion. Eventually I opted on not waiting for deleted VMs, not sure it gives us anything anyway.

Reply:
Nasty.  We should be taking that up with MSFT.
If we don't wait for VMs to delete we will unblock the next RP operation on the cluster straightaway and that could cause us quite confusing problems.

Reply:
@amanohar @sozercan @shrutir25

Reply:
We may need to bump the default context timeout:
```
INFO[2018-08-30T13:09:40+02:00] draining compute-000002                       resourceGroup=mkargakiosa
INFO[2018-08-30T13:09:42+02:00] deleting compute-000002                       resourceGroup=mkargakiosa
FATA[2018-08-30T13:24:43+02:00] Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded 
```